torch/jit/python_frontend.py

```diff
@@ -0,0 +1,285 @@
+               build_param_list(ctx, py_def.args),
+               returns,
+               build_stmt_list(ctx, body) + ret_body)
+
+
+def build_param_list(ctx, py_args):
+    assert py_args.vararg is None
+    assert py_args.kwarg is None
+    assert not py_args.defaults
+    if PY2:
+        # TODO: args are in py_args.args, but are expressions <sigh>
+        raise RuntimeError("PY2 not supported")
+    else:
+        assert not py_args.kwonlyargs
+        assert not py_args.kw_defaults
+        return [build_param(ctx, arg) for arg in py_args.args]
+
+
+def build_param(ctx, py_arg):
+    assert py_arg.annotation is None  # TODO: handle annotations to get types
+    name = py_arg.arg
+    r = ctx.make_range(py_arg.lineno, py_arg.col_offset, py_arg.col_offset + len(name))
+    return Param(TensorType(r), Ident(r, name))
+
+
+def build_stmt_list(ctx, py_stmt_list):
+    return [build_stmt(ctx, stmt) for stmt in py_stmt_list]
+
+
+class StmtBuilder(Builder):
+    augassign_map = {
+        ast.Add: '+',
+        ast.Sub: '-',
+        ast.Mult: '*',
+        ast.Div: '/',
+    }
+
+    @staticmethod
+    def build_Expr(ctx, stmt):
+        return ExprStmt(build_expr(ctx, stmt.value))
+
+    @staticmethod
+    def get_assign_ident(ctx, expr):
+        var = build_expr(ctx, expr)
+        if not isinstance(var, Var):
+            raise RuntimeError("the only expressions allowed on the left hand side of "
+                               "assignments are variable names")
+        return var.name()
+
+    @staticmethod
+    def build_Assign(ctx, stmt):
+        return Assign([StmtBuilder.get_assign_ident(ctx, e) for e in stmt.targets],
+                      '=',
+                      build_expr(ctx, stmt.value))
+
+    @staticmethod
+    def build_AugAssign(ctx, stmt):
+        op = type(stmt.op)
+        if op in StmtBuilder.augassign_map:
+            op_token = StmtBuilder.augassign_map[op]
+        else:
+            raise RuntimeError("unsupported kind of augumented assignment: " + op.__name__)
+        return Assign([StmtBuilder.get_assign_ident(ctx, stmt.target)],
+                      op_token,
+                      build_expr(ctx, stmt.value))
+
+    @staticmethod
+    def build_While(ctx, stmt):
+        if stmt.orelse:
+            raise RuntimeError("else branches of while loops aren't supported")
+        r = ctx.make_range(stmt.lineno, stmt.col_offset, stmt.col_offset + len("while"))
+        return While(r, build_expr(ctx, stmt.test), [build_stmt(ctx, s) for s in stmt.body])
+
+    @staticmethod
+    def build_If(ctx, stmt):
+        r = ctx.make_range(stmt.lineno, stmt.col_offset, stmt.col_offset + len("if"))
+        return If(r, build_expr(ctx, stmt.test),
+                  [build_stmt(ctx, s) for s in stmt.body],
+                  [build_stmt(ctx, s) for s in stmt.orelse])
+
+
+class ExprBuilder(Builder):
+    _MethodRef = namedtuple('MethodRef', ['self', 'name'])
+    binop_map = {
+        ast.Add: partial(BinOp, '+'),
```

Comment:
Why use `partial` here and a tuple in `unop_map`? I find that inconsistency confusing. I'd prefer avoiding partial entirely since it relys on the reader knowing what it is.

Reply:
Mostly because we don't really care about having accurate ranges for binary operations, because they are written infix and unification of subtree ranges will include them anyway. Unary operators are prefixing the expression, so I need to take the `col_offset:col_offset+len(token)` as an "extra" range argument to `UnaryOp::create`. I think I could have changed it to be `col_offset:col_offset+1` and have the unification still work out alright, but it seems a bit hacky.

If the reader doesn't know `partial`s they should read up on them. They're such a useful tool 😉 

I'm ok with changing all those maps to contain only the tokens, and then call `BinOp` and `UnaryOp` in the respective methods if you want them to be consistent.